packages/cli/src/lib/implementation/core-config.middleware.integration.test.ts

```diff
@@ -35,3 +50,80 @@ describe('coreConfigMiddleware', () => {
     ).rejects.toThrow(/Provided path .* is not valid./);
   });
 });
+
+describe('cliWithConfigOptionsAndMiddleware', () => {
+  const cliWithConfigOptionsAndMiddleware = (
+    cliObj: Partial<CoreConfigCliOptions>,
+  ) =>
+    yargsCli<CoreConfig>(objectToCliArgs(cliObj), {
+      options: {
+        ...yargsCoreConfigOptionsDefinition(),
+      },
+      middlewares: [{ middlewareFunction: coreConfigMiddleware }],
+    });
+  const configPath = (kind?: CoreConfigName) => {
+    const templateKind = kind ? `${kind}.` : '';
+    return join(configDirPath, `code-pushup.${templateKind}config.ts`);
+  };
+  const cliPersistOptions: PersistConfigCliOptions = {
+    'persist.outputDir': 'tmp-cli',
+    'persist.format': 'md',
+    'persist.filename': 'cli-report',
+  };
+  const cliResultPersistOptions: PersistConfig = {
+    outputDir: cliPersistOptions['persist.outputDir'],
+    format: [cliPersistOptions['persist.format']],
+    filename: cliPersistOptions['persist.filename'],
+  };
+  const rcResultPersistOptions: PersistConfig = {
+    outputDir: 'rc-tmp',
+    format: ['json', 'md'],
+    filename: 'rc-report',
+  };
+
+  it('should take default values for persist when no argument is given in rc or over the cli', async () => {
+    const expectedObject = {
+      outputDir: PERSIST_OUTPUT_DIR,
+      format: PERSIST_FORMAT,
+      filename: PERSIST_FILENAME,
+    };
+    const argv = await cliWithConfigOptionsAndMiddleware({
+      config: configPath('minimal'),
+    }).parseAsync();
+
+    expect(argv?.persist).toEqual(expect.objectContaining(expectedObject));
+  });
+
+  it('should take values for persist when provided over the cli', async () => {
+    const expectedObject = cliResultPersistOptions;
+    const argv = await cliWithConfigOptionsAndMiddleware({
+      ...cliPersistOptions,
+      config: configPath('minimal'),
+    }).parseAsync();
+
+    expect(argv?.persist).toEqual(expect.objectContaining(expectedObject));
+  });
+
+  it('should take values for persist when provided over the rc', async () => {
+    const expectedObject = rcResultPersistOptions;
+    const argv = await cliWithConfigOptionsAndMiddleware({
+      config: configPath('persist'),
+    }).parseAsync();
+
+    expect(argv?.persist).toEqual(expect.objectContaining(expectedObject));
+  });
+
+  it('should take values for persist when provided over both, the rc config and the cli', async () => {
```

Comment:
It would be valuable to highlight here that if all three types of sources exist (default, config, CLI), the CLI takes precedence.

Reply:
added the test